core/client/hdfs/src/main/java/alluxio/hadoop/HdfsFileInputStream.java

```diff
@@ -39,7 +43,7 @@
  */
 @NotThreadSafe
 public class HdfsFileInputStream extends InputStream implements Seekable, PositionedReadable,
-    ByteBufferReadable {
+    ByteBufferReadable, StreamCapabilities, CanUnbuffer {
```

Comment:
do we know staring from which hadoop version , these interfaces are introduced?

Reply:
https://issues.apache.org/jira/browse/HDFS-7694 
This ticket show that it string from 2.6.4

Reply:
then one problem will be: if we build Alluxio src code with Hadoop version < 2.6.4, the build will fail.
is it possible that we add `hasCapability` and `unbuffer` here, but without claiming implementing
interfaces `StreamCapabilities`, `CanUnbuffer`? will that work for Impala? e.g., does Impala checks if the InputStream extends these interfaces explicitly?

Reply:
@apc999 Ok, I will ask impala contributor who ask for this feature if can not check it

Reply:
as discussed offline, we can follow the same convention as introducing `ByteBufferReadable`

Reply:
@ggezer  in short, this PR is not introducing any new change in backward compatibility when building Alluxio.

This is because `org.apache.hadoop.fs.ByteBufferReadable` (introduced a few years back into `HdfsFileInputStream`) already breaks the build. As a result, if someone wants to build the Alluxio with north-bound HDFS client to be 2.2 or 1.x, they will already fail. Actually the two interfaces `StreamCapabilities`, `CanUnbuffer` is introduced to Hadoop earlier than `ByteBufferReadable`, so this PR is not making this situation  worse. 

Thus I think this change shall be fine, given we don't even notice this break in the last few years --- largely meaning no one attempts to build Alluxio north bound to be HDFS 1.x or 2.2. So we shall be good.